### PR TITLE
Add dictionary merge solution and related test

### LINKED ISSUE #33: Dictionary-Merging-with-Conflict-Resolution
Challenge: Merge two dictionaries with conflict resolution.
Description: A function that merges two dictionaries. If both dictionaries have the same key, the value in the second dictionary (dict_b) should overwrite the value in the first dictionary (dict_a). Additionally, implement conflict resolution where, in case of a conflict, the larger value (based on comparison) is kept. If the value is not comparable, keep the value from the first dictionary.

### PR DESCRIPTION
**Pull request by Frank2446-dotcom**
---
name: solution review 
about: A template PR for code review with a checklist
---

### Files

- [x] The file name describes the function's behavior
- [x] There is a module docstring in the function file
- [x] The test file's name matches the function file name -
  `/tests/test_file_name.py`
- [ ] There is a module docstring in the tests file

### Unit Tests

- [ ] The test class has a helpful name in PascalCase
- [ ] The test class has a docstring
- Every unit test has
  - [x] A helpful name
  - [ ] A clear docstring
  - [x] Only one assertion
  - [x] There is no logic in the unit test
- [x] All tests pass
- [ ] There are tests for defensive assertions
- [x] There are tests for boundary cases

### Function Docstring

- [x] The function's behavior is described
- The function's arguments are described:
  - [x] Type
  - [x] Purpose
  - [x] Other assumptions (eg. if it's a number, what's the expected range?)
- The return value is described
  - [x] Type
  - [x] Other assumptions are documented
- The defensive assertions are documented using `Raises:`
  - [ ] Each assumption about an argument is checked with an assertion
  - [ ] Each assertion checks for _only one_ assumption about the argument
- [ ] Include 3 or more (passing!) doctests

### The Function

- [x] The function's name describes it's behavior
- [x] The function's name matches the file name
- [x] The function has correct type annotations
- [x] The function is not called in the function file

### Do's

- [x] Variable names help to understand the strategy
- [x] Any comments are clear and describe the strategy
- [ ] Lines of code are spaced to help show different stages of the strategy

### Don'ts

- [x] The function's strategy _is not_ described in the documentation
- [x] Comments explain the _strategy_, **not** the _implementation_
- [x] The function _does not_ have more comments than code
  - If it does, consider finding a new strategy or a simpler implementation

## Implementation

- [x] The code passes the formatting checks
- [x] The code passes all Ruff linting checks
- [x] The code has no (reasonable) Pylint errors
  - In code review, you can decide when fixing a Pylint error is helpful and
    when it's too restricting.
- [x] Variables are named with snake_case
- [x] Variable names are clear and helpful
- [x] The code follows the strategy as simply as possible
- [x] The implementation is as simple as possible given the strategy
- [x] There are no commented lines of code
- [ ] There are no `print` statements anywhere
- [ ] The code includes defensive assertions
- [ ] Defensive assertions include as little logic as possible

### CODE REVIEW

File: solutions/tests/test_merge_dictionaries.py [hunk 1]
```diff
@@ -0,0 +1,52 @@
+# solutions/tests/test_merge_dictionaries.py
+
+import unittest
+from solutions.merge_dictionaries import merge_dictionaries
+
+
```
Comment (TibyanKhalid) [inline solutions/tests/test_merge_dictionaries.py:16]: You should add a module docstring to the Tests file

File: solutions/tests/test_merge_dictionaries.py [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+class TestMergeDictionaries(unittest.TestCase):
```
Comment (TibyanKhalid) [inline solutions/tests/test_merge_dictionaries.py:17]: The test class should have the name in PascalCase
You should add a docstring to the test class

File: solutions/tests/test_merge_dictionaries.py [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+    def test_no_conflicts(self):
```
Comment (TibyanKhalid) [inline solutions/tests/test_merge_dictionaries.py:20]: You should add docstrings to each unit test that describes what the test is checking

File: solutions/tests/test_merge_dictionaries.py [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+        dict2 = {}
+        expected = {"a": 1, "b": 2}
+        self.assertEqual(merge_dictionaries(dict1, dict2), expected)
+
+    def test_only_conflicts(self):
+        dict1 = {"a": 1}
+        dict2 = {"a": 2}
+        expected = {"a": 2}
+        self.assertEqual(merge_dictionaries(dict1, dict2), expected)
+
```
Comment (TibyanKhalid) [inline solutions/tests/test_merge_dictionaries.py:68]: You should add defensive assertions tests ( Tests that focus on verifying that a function has internal checks (like assertions) to catch invalid states or inputs )

File: solutions/merge_dictionaries.py [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+
+    return merged
+
+
+if __name__ == "__main__":
+    # Example usage
+    dict_a = {"x": 1, "y": 2}
+    dict_b = {"y": 3, "z": 4}
+
+    print("Default merge (dict_b overwrites):")
```
Comment (TibyanKhalid) [inline solutions/merge_dictionaries.py]: You should NOT have print statements anywhere

File: solutions/merge_dictionaries.py [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+            two arguments (value1, value2) and returns the resolved value.
+
+    Returns:
+        dict: A merged dictionary.
+
+    Example:
+        >>> merge_dictionaries({'a': 1}, {'a': 2, 'b': 3})
+        {'a': 2, 'b': 3}
+        >>> merge_dictionaries({'a': 1}, {'a': 2, 'b': 3}, max)
+        {'a': 2, 'b': 3}
```
Comment (TibyanKhalid) [inline solutions/merge_dictionaries.py:37]: You should add more doctests ( you only have 2, there should be at least 3 )

File: solutions/merge_dictionaries.py [hunk 7]
```diff
@@ -0,0 +1,49 @@
...
+    """
```
Comment (TibyanKhalid) [inline solutions/merge_dictionaries.py:40]: You should add defensive assertions and they should include as little logic as possible
You should document defensive assertions using Raises:
1. Each assumption about an argument must be checked with an assertion 
2. Each assertion should check for only 1 assumption about the argument

**Review by TibyanKhalid — CHANGES_REQUESTED** [review]
Hey Frank, these are the stuff that were missing from the challenge and the tests file, they are the boxes that I didn’t tick in the checklist as well

**Comment by TibyanKhalid** [comment]
I can see that you added all the changes and edits required, good job!! It's ready to merge now
